### PR TITLE
[SPARK-14786] Remove hive-cli dependency from hive subproject

### PR DESCRIPTION
**Pull request by JoshRosen**
The `hive` subproject currently depends on `hive-cli` in order to perform a check to see whether a `SessionState` is an instance of `org.apache.hadoop.hive.cli.CliSessionState` (see #9589). The introduction of this `hive-cli` dependency has caused problems for users whose Hive metastore JAR classpaths don't include the `hive-cli` classes (such as in #11495).

This patch removes this dependency on `hive-cli` and replaces the `isInstanceOf` check by reflection. I added a Maven Enforcer rule to ban `hive-cli` from the `hive` subproject in order to make sure that this dependency is not accidentally reintroduced.

/cc @rxin @yhuai @adrian-wang @preecet

### CODE REVIEW

File: sql/hive/pom.xml [hunk 1]
```diff
@@ -225,6 +221,27 @@
...
+        <executions>
+          <execution>
+            <id>enforce-versions</id>
+            <goals>
+              <goal>enforce</goal>
+            </goals>
+            <configuration>
+              <rules>
+                <bannedDependencies>
+                  <excludes combine.children="append">
```
Comment (JoshRosen) [inline sql/hive/pom.xml]: To verify that this does the correct thing, see

```
mvn -Phive -pl :spark-hive_2.11 help:effective-pom
```

**Comment by rxin** [comment]
LGTM pending Jenkins.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala [hunk 2]
```diff
@@ -110,10 +109,20 @@ private[hive] class HiveClientImpl(
       }
     }
 
+    def isCliSessionState(state: SessionState): Boolean = {
+      var temp: Class[_] = state.getClass
```
Comment (JoshRosen) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: In a test, this ended up failing with an NPE so I guess that the comment on line 114 / 123 is incorrect :(

**Comment by adrian-wang** [comment]
LGTM

**Comment by JoshRosen** [comment]
Jenkins, retest this please.

**Comment by rxin** [comment]
Merging in master.

**Pull request closed without merging**